### PR TITLE
Add support for setting socket max num tries and reconnect timeout 

### LINKED ISSUE #140: TCPSocket loops forever - add timeout or max num tries
In TCPSocket::setup we are looping forever until connected. It would be great to add a timeout or max_num_tries as an argument.

### PR DESCRIPTION
**Pull request by michaelaeriksen**
Fixes #140

### CODE REVIEW

**Comment by ahatstat** [comment]
This would be helpful to have.  Is there a plan to merge this PR?

File: CMakeLists.txt [hunk 1]
```diff
@@ -49,7 +49,7 @@ add_library(urcl SHARED
 )
 add_library(ur_client_library::urcl ALIAS urcl)
 target_compile_options(urcl PRIVATE -Wall -Wextra -Wno-unused-parameter)
-target_compile_options(urcl PUBLIC ${CXX17_FLAG})
+target_compile_options(urcl PRIVATE ${CXX17_FLAG})
```
Comment (fmauch) [inline CMakeLists.txt:52]: What was your motivation for that switch? In my point of view it does make sense to export this to downstream packages.

A better approach would be to expose [`target_compile_features`](https://cmake.org/cmake/help/latest/manual/cmake-compile-features.7.html#requiring-language-standards) but I don't really see this relevant in the scope of this PR.

File: include/ur_client_library/comm/tcp_socket.h [hunk 2]
```diff
@@ -50,7 +50,7 @@ class TCPSocket
 private:
   std::atomic<int> socket_fd_;
   std::atomic<SocketState> state_;
-  std::chrono::seconds reconnection_time_;
+  std::chrono::milliseconds reconnection_time_;
```
Comment (fmauch) [inline include/ur_client_library/comm/tcp_socket.h:53]: Why has this been changed to milliseconds?

File: include/ur_client_library/comm/producer.h [hunk 3]
```diff
@@ -60,13 +60,14 @@ class URProducer : public IProducer<T>
   /*!
    * \brief Triggers the stream to connect to the robot.
    */
-  void setupProducer() override
+  void setupProducer(size_t max_num_tries = 0, std::chrono::milliseconds reconnection_time = std::chrono::seconds(10)) override
```
Comment (fmauch) [inline include/ur_client_library/comm/producer.h:63]: We try to document **all** parameters to public methods.

File: include/ur_client_library/comm/pipeline.h [hunk 4]
```diff
@@ -172,7 +172,7 @@ class IProducer
   /*!
    * \brief Set-up functionality of the producers.
    */
-  virtual void setupProducer()
+  virtual void setupProducer(size_t max_num_tries = 0, std::chrono::milliseconds reconnection_time = std::chrono::seconds(10))
```
Comment (fmauch) [inline include/ur_client_library/comm/pipeline.h:175]: We try to document **all** parameters to public methods.

File: include/ur_client_library/comm/pipeline.h [hunk 5]
```diff
@@ -271,9 +271,9 @@ class Pipeline
     stop();
   }
 
-  void init()
+  void init(size_t max_num_tries = 0, std::chrono::milliseconds reconnection_time = std::chrono::seconds(10))
```
Comment (fmauch) [inline include/ur_client_library/comm/pipeline.h:274]: We try to document **all** parameters to public methods.

File: include/ur_client_library/ur/dashboard_client.h [hunk 6]
```diff
@@ -55,15 +55,15 @@ class DashboardClient : public comm::TCPSocket
...
+  static constexpr int DASHBOARD_SERVER_PORT = 29999;
 
   /*!
    * \brief Opens a connection to the dashboard server on the host as specified in the constructor.
    *
    * \returns True on successful connection, false otherwise.
    */
-  bool connect();
-
+  bool connect(size_t max_num_tries = 0, std::chrono::milliseconds reconnection_time = std::chrono::seconds(10));
```
Comment (fmauch) [inline include/ur_client_library/ur/dashboard_client.h:65]: We try to document **all** parameters to public methods.

**Review by fmauch — CHANGES_REQUESTED** [review]
Thank you for the contribution @michaelaeriksen!

I like the idea in general. There are a couple of docstrings missing and a couple of changes seem not necessarily be connected to the actual PR topic, please see my comments.

**Pull request closed without merging**